pkg/writer/serializer_cdx.go

```diff
@@ -13,9 +13,19 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-type SerializerCDX struct{}
+type SerializerCDX struct {
+	rootDict       map[string]struct{}
+	addedDict      map[string]struct{}
+	componentsDict map[string]*cdx.Component
+	reflessList    []*cdx.Component
```

Comment:
I think we can drop these in favor of a simpler serialization state object we can pass around to the internal methods. To simplify, we can drop:

- `rootDict` (see comment below)
- `reflessList` This one we should drop because nodes without an ID is something we cannot support (I'll detail more in a separate issue).

This means that we can have a single state object like this:
```golang
type cdxSerializationState struct {
	addedDict      map[string]struct{}
	componentsDict map[string]*cdx.Component
}
```

We could drop the added index too if we handle dependencies first. 

We can then pass around a pointer to the state to each internal function in a context or directly like this:

```go
  serializer.dependencies(state cdxSerializationState, *bom *sbom.Document)
```

The benefit of doing it like this is that we can handle as many serializations in parallel as we want without the need for mutexes, etc

Reply:
1) Removing refless (placeholder for generating the ref for now)
2) removing rootDict (not used).

Q: Wouldn't two parallel serializations have two instances of the  SerializerCDX?

Reply:
I added `SerializerCDXState` as you suggested.
i was thinking maybe we like to use the context library that may be a good fit,
if you think its not needed i can remove this layer and simply provide the state pointer as you suggested.